win-odbc64/odbcclient/drvr35/cdiag.cpp

```diff
@@ -389,8 +389,8 @@ SQLRETURN CDiagRec::GetDiagRec(SQLSMALLINT	RecNumber,
 					translateLengthMax/2, (int *)&translateLength, (char *)errorMsg)) != SQL_SUCCESS )
 				rc = SQL_SUCCESS_WITH_INFO; //ERROR;
 
-			strLen = translateLength; 
-			((wchar_t *)MessageText)[strLen] = L'\0' ; 
+			((wchar_t *)MessageText)[translateLength] = L'\0';
+			strLen = tmpStrLen;

```

Comment:
As per the ODBC specification, if the MessageText is NULL, the TextLengthPtr should give the caller the number of characters that will be returned. 

Usually, the ODBC applications are expected to get the number of characters needed for the error message by passing in MessageText as NULL, then allocate that many characters in their space and then request for the error message by passing the pointer to the buffer allocated as the MessageText.  

It looks like the logic of returning the actual number of characters needed when MessageText is NULL is missing